[mkdocs] feat: send messages tutorial (#1738)
symbol/symbol: mkdocs/pages/en/devbook/transactions/messages.md

---
title: Messages
---

# Sending Messages with Transfer Transactions

<Transfer transactions:> can include an optional message field, which allows attaching up to 1,024 bytes of data to the
transaction.
Messages can be sent as plain text or encrypted using the recipient's public key, ensuring only the intended recipient
can read them.

This tutorial shows how to send both plain and encrypted messages, and demonstrates how to decrypt received messages.

## Prerequisites

If you have not done so already, start with the [Hello World](../start/hello-world.md) tutorial to make sure your
development environment is set up correctly.

> Additionally, check the [Transfer transaction](transfer.md) tutorial to understand how fee

Reviewer: ```suggestion
Additionally, check the [Transfer transaction](./transfer.md) tutorial to understand how fee
```

It's not critical, but in the past I've used a plugin that automatically searched for files if they had not path.
I'm not using it now, but just in case I'll ever bring it back, always include the path in file names, even if it's just `./`

calculation, network time, and transaction confirmation works.

## Full Code

{% import 'tutorial.jinja2' as tutorial with context %}

{{ tutorial.code_full('devbook/transactions/messages', ['py', 'js']) }}

## Code Explanation

### Setting Up Accounts

{{ tutorial.code_snippet(['py:14:43', 'js:10:32']) }}

The tutorial uses two accounts: a sender and a recipient.
The code reads the sender's private key from the `PRIVATE_KEY` environment variable and the recipient's public key from
the `RECIPIENT_PUBLIC_KEY` environment variable.
Both variables default to test keys if not set.

In a real application, you typically have the recipient's address rather than their public key.
To encrypt messages, you need to obtain their public key by querying the network using the <get:/accounts/{accountId}>
endpoint.

> The public key is only available for accounts that have previously sent transactions.

Reviewer: The reader and the docs might disagree on what "a real application" is :)
Let's stick to the facts. Also, this looks like a nice opportunity for an info box. How about:
```suggestion
!!! note
    A recipient's <address:> is sufficient to send a message, but encryption requires their <public key:>.
    When only the address is known, the public key can be retrieved from the network using the <get:/accounts/{accountId}> endpoint.
    An account's public key becomes available only after it has broadcast at least one transaction.
```

### Sending a Plain Text Message

> {{ tutorial.code_snippet(['py:45:108', 'js:34:91']) }}

Reviewer: This snippet is way too long. We don't need to show all the code, just the parts relevant to the topic of the tutorial.

You can leave out the parts about fetching network time and fees, and in the text you can send readers to the Transfer tutorial.
Same thing with the part about announcing the tx, that's also explained in the Transfer tutorial.
So the snippet should only be lines 57 to 74 (in the JS file)

To attach a plain text message to a transfer transaction, encode the message as bytes and include it in the `message`

> field.

Reviewer: I think this is a good place to talk about encodings, because we just told the readers to "encode" the message, but we didn't say what code they should use.

How about adding the paragraph from the Textbook:
> The Symbol protocol does not define a standard encoding for messages.
> It is up to the sender and recipient to agree on a format.
> By convention, most wallets and applications assume UTF-8 encoding.

The transaction follows the same structure as a [basic transfer transaction](transfer.md), with the addition of the
message field:

> * **Message format:** Plain text messages are encoded directly as UTF-8 bytes.

Reviewer: Now that we've talked about encodings, this line is no longer necessary (and it's not true, UTF-8 is a convention but anything else can be used).

* **Maximum size:** Messages cannot exceed 1,024 bytes. The network will reject transactions with larger messages.

    !!! tip "Handling larger data"

        For applications requiring more than 1,024 bytes of data, consider storing the data off-chain and including a
        hash or reference in the message field.

>         This approach maintains data integrity verification while keeping transaction sizes manageable.

Reviewer: This is a good addition!
Just two comments:
* I'd say "include a hash AND a reference", you need one to retrieve the message and the other to verify its integrity.
* I'd mention that you can also keep everything on-chain by using multiple transactions and an aggregate.

* **Network visibility:** Plain text messages are publicly visible on the blockchain.

>     Anyone can read them by querying the transaction.

Reviewer: It's a nitpick, but encrypted messages are also publicly visible... just not decipherable. So I'd rephrase slightly:
```suggestion
* **Network visibility:** All messages are publicly visible on the blockchain, so unless they are encrypted,
    anyone can read them by querying the transaction.
```

This example sends 1 XYM along with the message `"Hello, Symbol!"` to demonstrate that messages can be combined with
mosaic transfers.

!!! info "Messages without mosaics"

>     You can send a message without transferring any mosaics by providing an empty mosaics array.

Reviewer: Let's reverse this. We don't send any mosaic in the code, for simplicity, and just say here that you can combine mosaic transfers and messages if you want.

### Sending an Encrypted Message

> {{ tutorial.code_snippet(['py:110:167', 'js:93:142']) }}

Reviewer: Again, too long. Skip the fetching of the network time (which I'd rather remove, as commented above) and the announcement of the tx.

> Encrypted messages provide confidentiality by encrypting the message content using a shared secret derived from the

Reviewer: ```suggestion
Encrypted messages provide confidentiality by protecting the message content using a shared secret derived from the
```

Minor tweak to avoid repetition. The paragraph is good!

sender's private key and the recipient's public key.
Both the sender and recipient can decrypt the message using their own private key and the other party's public key.

> The encryption process uses the <py:MessageEncoder> class:

Reviewer: Ugh, this should really be `<dy:MessageEncoder>` but I see the JS link was broken.
I pushed a commit to `new-docs` that fixes it, you can go back to using `dy:` everywhere :)

1. **Create a MessageEncoder** with the sender's key pair.
2. **Encode the message** using the recipient's public key and the message bytes with <py:MessageEncoder.encode>.
3. **Attach the encrypted payload** to the transaction's `message` field.

By convention, encrypted messages begin with a `0x01` byte to indicate encryption, followed by the encrypted content.
This reduces the effective message size to 1,023 bytes.

!!! info "Message encryption is a convention, not a protocol feature"

    The Symbol protocol treats the message field as a byte array (max 1,024 bytes) without validating its content.

    By convention, applications use a prefix byte to indicate message encoding:

    * `0x00` - Unencrypted message
    * `0x01` - Encrypted message (using Bouncy Castle's AES block cipher in CBC mode)

    The SDK's `MessageEncoder` handles this convention, allowing both sender and recipient to decrypt messages using
    each other's public keys.

>     See [Transfer Transactions](../../textbook/transfer_transactions.md#optional-message) in the textbook for details.

Reviewer: This is already the content of the info box in the Textbook, no need to repeat it.

How about something like:
```suggestion
Just as the Symbol protocol does not define a standard for message encoding, it also does not define one for encryption.
Whether a message is encrypted must be agreed upon in advance between the sender and the recipient.

However, some conventions have been adopted by applications over time.
For more details, see the note on [Optional Messages](../../textbook/transfer_transactions.md#optional-message) in the Textbook.
```

### Decrypting a Received Message

{{ tutorial.code_snippet(['py:169:188', 'js:144:164']) }}

> To decrypt a message, create a <py:MessageEncoder> with your key pair and call <py:MessageEncoder.try_decode> method

Reviewer: ```suggestion
To decrypt a message, create a <py:MessageEncoder> with one key pair (either the sender's or the recipient's) and call the <py:MessageEncoder.try_decode> method
```

To make it explicit that decoding can be used in two different ways.

with the other party's public key and the encrypted payload.

> In this example, the sender decrypts using their own key pair and the recipient's public key to verify the encryption.

Reviewer: You'll have to adapt this to match the code.

The recipient would decrypt using their own key pair and the sender's public key (available in the transaction's
`signerPublicKey` field).

The method returns a tuple `(is_decoded, message)` indicating whether decryption was successful.

> If decryption succeeds, the message contains the original plaintext bytes, which you can decode as UTF-8.

Reviewer: ```suggestion
The method returns a tuple `(is_decoded, message)` indicating whether decryption was successful, and, if so, contains the original plaintext bytes, which still need to be decoded.
```

Reviewer: I think this comment still applies. It's weird to see that the method returns `(is_decoded, message)` but the text only says that it returns whether decryption was successful. Yes, `message` is explained in the next paragraph, but if people try to understand paragraph by paragraph, this is confusing.

Also, I don't want to say "which you can decode as UTF-8", because we already said that's a convention, maybe some app wants to use a different encoding.

For these reasons, I still prefer my suggestion.

If decryption fails, possible causes include:

* The message was encrypted for a different recipient.
* The message is corrupted or tampered with.
* The message is plain text, not encrypted.
* An incorrect public key was used for the other party.

## Output

The output shown below corresponds to a typical run of the program.

```text
--8<-- 'devbook/transactions/messages.log'
```

You can view the transactions on the [Symbol Testnet Explorer](https://testnet.symbol.fyi/) by searching for the
transaction hashes printed in the output.

When you view an encrypted message transaction in the explorer, the message field will appear as hexadecimal data

> beginning with `01`.

Reviewer: This does not work for symbol.fyi, you just get a random jumble, not a hex string.

The explorer cannot decrypt the message because it doesn't have access to the private keys.

## Conclusion

This tutorial showed how to:

| Step                                                           | Related documentation                     |
| -------------------------------------------------------------- | ----------------------------------------- |

> | [Set up accounts](#setting-up-accounts)                        |                                           |

Reviewer: This part was not related to the topic, not worth mentioning here.

I think I would only list:
- Convert text into UTF-8 bytes using TextEncoder
- Encrypt and Decrypt a message using MessageEncoder
- How to include a message in a Transfer Transactions using SymbolTransactionFactory

{{ tutorial.code_snippet(['py:48:74', 'js:44:62']) }}

To send a message, you need the sender's <private key:> and the recipient's <address:>.
To encrypt a message, you additionally need the recipient's <public key:>.

This tutorial uses two accounts (sender and recipient) to demonstrate both sending and receiving plain and encrypted
messages.
The snippet reads their private keys from the `SENDER_PRIVATE_KEY` and `RECIPIENT_PRIVATE_KEY` environment variables,
which default to test keys if not set.
The recipient's public key and address are derived from their private key.

!!! note "Retrieving public keys"

    When only the address is known, you can retrieve the public key from the network using the
    <get:/accounts/{accountId}> endpoint.
    An account's public key becomes available only after it has broadcast at least one transaction.

### Sending a Plain Text Message

{{ tutorial.code_snippet(['py:99:113', 'js:87:100']) }}

You can also combine mosaic transfers with messages by including both the `mosaics` and `message` fields in the
 transaction descriptor.

See the [Creating a Transfer Transaction](./transfer.md) tutorial for details on fetching network time, calculating

>  fees, and announcing transactions.

Reviewer: This is a good comment, but I would move it to the intro of the Code Explanation section, in case anybody is following the code and is wondering what this is, before reaching this section.

You can say something like "the part about fetching network time and fees has been explained in the Transfer tutorial and is skipped here for brevity's sake."

**Message constraints:**

* **Maximum size:** 1,024 bytes (the network rejects larger messages).
* **Encoding:** UTF-8 by convention, though the protocol doesn't enforce a standard.
* **Privacy:** All messages are publicly visible on the blockchain unless encrypted.

!!! tip "Handling larger data"

    For applications requiring more than 1,024 bytes of data, common approaches include:

    * **On-chain storage:** Split the data across multiple transactions within an <aggregate transaction:>, allowing
        you to keep everything on the blockchain.
    * **Off-chain storage:** Store the data off-chain and include a hash and a reference in the message field.
        The hash verifies data integrity while the reference enables retrieval.

### Receiving a Plain Text Message

{{ tutorial.code_snippet(['py:138:151', 'js:120:130']) }}

> After announcing the transaction, the `wait_for_confirmation` helper function polls the

Reviewer: If you rename this function remember to search for the old name everywhere :)

> `/transactions/confirmed/{hash}` endpoint until the transaction is confirmed.

Reviewer: ```suggestion
<get:/transactions/confirmed/{transactionId}> endpoint until the transaction is confirmed.
```

The confirmed transaction contains the message as hex-encoded bytes.
To retrieve the original message, it converts the hex string to bytes and decodes it as UTF-8.

### Sending an Encrypted Message

{{ tutorial.code_snippet(['py:154:180', 'js:133:157']) }}

Encrypted messages provide confidentiality by protecting the message content using a shared secret derived from the
sender's private key and the recipient's public key.
Both the sender and recipient can decrypt the message using their own private key and the other party's public key.

The <dy:MessageEncoder> class handles message encryption:

1. A <dy:MessageEncoder> is created with the sender's key pair.
2. The message is encoded using the recipient's public key and the message bytes with <dy:MessageEncoder.encode>.
3. The encrypted payload is attached to the transaction's `message` field.

The transaction is then signed and announced following the same process as in
[Creating a Transfer Transaction](./transfer.md).

!!! note "Message encryption is a convention"

    The Symbol protocol does not define a standard for message encryption.

>     Sender and recipient must agree in advance on whether messages are encrypted.

Reviewer: ```suggestion
    Sender and recipient must agree in advance on whether messages are encrypted and the cipher used.
```

    Because the encryption uses a shared secret derived from both key pairs, the sender can also decrypt the message
    using their own private key and the recipient's public key.
    This allows both parties to verify the message content after it has been published on the blockchain.

If decryption fails, possible causes include:

* The message was encrypted for a different recipient.
* The message is corrupted or tampered with.
* The message is plain text, not encrypted.
* An incorrect public key was used for the other party.

## Output

The output shown below corresponds to a typical run of the program.

```text
--8<-- 'devbook/transactions/messages.log'
```

You can view the transactions on the [Symbol Testnet Explorer](https://testnet.symbol.fyi/) by searching for the
transaction hashes printed in the output.

The explorer cannot decrypt encrypted messages because it doesn't have access to the private keys.

## Conclusion

This tutorial showed how to:

| Step                                                           | Related documentation                       |
| -------------------------------------------------------------- | ------------------------------------------- |

> | [Convert text into UTF-8 bytes](#sending-a-plain-text-message) | `TextEncoder`                               |

Reviewer: Maybe point out that `TextEncoder` (in JS) and the string `encode` and `decode` methods (in Python) are system methods and not part of the Symbol SDK.

| [Encrypt and decrypt a message](#sending-an-encrypted-message) | <dy:MessageEncoder>                         |
| [Include a message in a Transfer Transaction](#sending-a-plain-text-message) | <dy:SymbolTransactionFactory> |

> 

Reviewer: Extra blank line.
